### DISCUSSION TITLE
Does sanitize-html convert "messy" HTML (which you find across the web) into standard HTML?

### DISCUSSION THREAD

**Discussion by lancejpollard** [apostrophecms/sanitize-html#641] [Q&A] [answered]
Years ago I read that the chrome browser could parse any HTML, even when you had self-closing tags on non-self-closable elements like `<div />` and other "messy" stuff. I would assume it can even recover from things like `<div <p>hello world</p</div>`, but not totally sure.

Does `sanitize-html` handle this sort of stuff? Basically I'm wondering if someone inputs some arbitrary webpage HTML, will it return more "standardized" HTML as output, like what Chrome would see? In this way, I don't have to go down the road of using `html-parser2` or `parse5` to parse messy HTML into an AST, then convert the AST back to standards compliant HTML.

This way, when converting HTML to PDF with pandoc (it successfully creates a malformed PDF with malformed HTML), I could first "sanitize" the HTML, to get standards-compliant HTML, and then render that to PDF. That would fix that problem, amongst other things.

If not, what does `sanitize-html` do differently?

**Comment by boutell** [comment, marked as answer]
sanitize-html uses htmlparser2 to parse the document, so you're going to get htmlparser2's interpretation of the document, for good or ill (generally for good in my experience, I'm not casting shade).

I think that answers the question, but if you have further questions about it you can check into the htmlparser2 documentation.
